### PR TITLE
[TO BE DISCUSSED] make generic/closure ByteBuffer methods inlineable

### PR DESCRIPTION
**Pull request by weissi**
I'm not 100% sure I did exactly the right thing and if we should to this. Paging @airspeedswift and @moiseev .

Motivation:

Generics and closures across modules can be problematic as the cause
allocations. ByteBuffer is one of NIO's core data types and therefore
should be well optimised. Generic/closure taking methods so far were a
bit of a problem. And with
[SE-0193](https://github.com/apple/swift-evolution/blob/master/proposals/0193-cross-module-inlining-and-specialization.md)
accepted, there's not much reason not to use those features even though they
are (at the moment) `@_`.

For the 1000 requests over 1 connection allocation benchmark we go from

    "total_allocations": 325378

to

    "total_allocations": 321366

so we lose 4 allocations per HTTP request/response pair, not super
impressive but not nothing either.

Modifications:

Make generic/closure taking methods on ByteBuffer `@_inlineable`

Result:

Fewer allocations

### CODE REVIEW

**Comment by weissi** [comment]
on Linux with Swift 4.0.3 we go from

    "total_allocations": 322051

to

    "total_allocations": 318047

so also lost 4 allocations.

**Comment by weissi** [comment]
@normanmaurer the actual name for `@_versioned` will (AFAIK) be `@usableFromInline`, another discussed name was `@abiPublic`. And yes, basically it means that a function (even if `internal`) can be used from other modules if inlined.

That change does make `ByteBuffer`s internal public ABI (not API) but that's okay as I don't foresee any `ByteBuffer` layout changes before we go SemVer major++ and also on the server we don't care about binary frameworks so YOLO.

**Comment by moiseev** [comment]
`@_versioned` things can also be `@_inlineable`. So you might squeeze some more improvement from it.

File: Sources/NIO/ByteBuffer-core.swift [hunk 1]
```diff
@@ -72,15 +72,12 @@ public struct ByteBufferAllocator {
...
-
-private func toCapacity(_ value: Int) -> Capacity {
-    return Capacity(truncatingIfNeeded: value)
+@_versioned func toCapacity(_ value: Int) -> ByteBuffer.Capacity {
+    return ByteBuffer.Capacity(truncatingIfNeeded: value)
 }
 
-private func toIndex(_ value: Int) -> Index {
-    return Index(truncatingIfNeeded: value)
+@_versioned func toIndex(_ value: Int) -> ByteBuffer.Index {
```
Comment (Lukasa) [inline Sources/NIO/ByteBuffer-core.swift]: Please do make the `@_versioned` properties underscore-prefixed.

**Comment by weissi** [comment]
thanks @moiseev . Looks better now?

**Comment by moiseev** [comment]
@Lukasa the changes have landed in master. So master branch of Swift right now supports both the old and the new spelling. There are no concrete plans to port it to the 4.2 branch right away. So in order to be compatible, for now, it is safe to stick to the old form of these attributes.

**Comment by moiseev** [comment]
@weissi It's not a matter of how it looks =) I commented primarily because it might enable some other optimizations.

**Comment by Lukasa** [comment]
@moiseev Sorry, I should be clearer. Is there any plan to *drop* these old names, e.g. in Swift 4.2? I'm very reluctant to ship code that we have reason to believe will break unnecessarily in a future release if we can possibly avoid it.

**Comment by weissi** [comment]
@Lukasa I think what @moiseev is saying that 4.2 will _only_ have the old names. Swift 5 will probably have old + new.

**Comment by normanmaurer** [comment]
@Lukasa WDYT? I think its "ok" to merge

**Comment by Lukasa** [comment]
I don’t think @moiseev has fully answered my question, so I’m still a bit nervous about merging this.

**Comment by weissi** [comment]
@Lukasa I think we're fine: https://twitter.com/slava_pestov/status/982997370659983360 . Slava implemented this AFAIK

**Comment by Lukasa** [comment]
Awesome, thanks. Lemme re-review with that in mind.

**Pull request merged**